ext/semian/sliding_window.c

```diff
@@ -0,0 +1,212 @@
+#include "sliding_window.h"
+
+#include <errno.h>
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include "util.h"
+
+static const rb_data_type_t semian_simple_sliding_window_type;
+
+static semian_simple_sliding_window_shared_t* get_window(uint64_t key) {
+  const int permissions = 0664;
+  int shmid = shmget(key, sizeof(int), IPC_CREAT | permissions);
+  if (shmid == -1) {
+    rb_raise(rb_eArgError, "could not create shared memory (%s)", strerror(errno));
+  }
+
+  void *val = shmat(shmid, NULL, 0);
+  if (val == (void*)-1) {
+    rb_raise(rb_eArgError, "could not get shared memory (%s)", strerror(errno));
+  }
+
+  return (semian_simple_sliding_window_shared_t*)val;
+}
+
+static int check_max_size_arg(VALUE max_size)
+{
+  int retval = -1;
+  switch (TYPE(max_size)) {
+  case T_NIL:
+    retval = SLIDING_WINDOW_MAX_SIZE; break;
+  case T_FLOAT:
+    rb_warn("semian sliding window max_size is a float, converting to fixnum");
+    retval = (int)(RFLOAT_VALUE(max_size)); break;
+  default:
+    retval = RB_NUM2INT(max_size); break;
+  }
+
+  if (retval <= 0) {
+    rb_raise(rb_eArgError, "max_size must be greater than zero");
+  } else if (retval > SLIDING_WINDOW_MAX_SIZE) {
+    rb_raise(rb_eArgError, "max_size must be less than %d", SLIDING_WINDOW_MAX_SIZE);
```

Comment:
`less than or equal to %d` or `not be greater than %d`

Reply:
Done.